DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp

```diff
@@ -0,0 +1,436 @@
+
+  [[intel::ivdep(kCacheDepth)]]  // NO-FORMAT: Attribute
+  do {
+    bool data_valid = true;
+    BytePack<1> out_data;
+
+    // if we aren't currently reading from the history, read from input pipe
+    if (!reading_history) {
+      // if we aren't currently reading from the history buffers,
+      // then read from input pipe
+      auto pipe_data = InPipe::read(data_valid);
+
+      // check if the upstream kernel is done sending us data
+      done = pipe_data.flag && data_valid;
+
+      // grab the literal or the length and distance pair
+      unsigned short dist = pipe_data.data.distance;
+
+      // for the case of a literal, we will simply write it to the output
+      out_data.byte[0] = pipe_data.data.literal[0];
+      out_data.valid_count = ac_uint<1>(1);
+
+      // if we get a length distance pair we will read 'pipe_data.data.length'
+      // bytes starting at and offset of 'dist'
+      history_counter = pipe_data.data.length;
+      reading_history = pipe_data.data.is_copy && data_valid;
+      reading_history_next = history_counter > 1;
+
+      // initialize the read index
+      read_history_buffer_idx =
+          (history_buffer_idx - dist) & history_buffer_idx_mask;
+    }
+
+    if (reading_history) {
+      // read from the history buffer
+      out_data.byte[0] = history_buffer[read_history_buffer_idx];
+
+      // also check the cache to see if it is there
+#pragma unroll
+      for (int j = 0; j < kCacheDepth + 1; j++) {
+        if (history_buffer_cache_idx[j] == read_history_buffer_idx) {
+          out_data.byte[0] = history_buffer_cache_val[j];
+        }
+      }
+      out_data.valid_count = ac_uint<1>(1);
+
+      // update the history read index
+      read_history_buffer_idx =
+          (read_history_buffer_idx + ac_uint<1>(1)) & history_buffer_idx_mask;
+
+      // update whether we are still reading the history
+      reading_history = reading_history_next;
+      reading_history_next = history_counter > 2;
+      history_counter--;
+    }
+
+    if (!done && data_valid) {
+      // write to the most history buffer
+      history_buffer[history_buffer_idx] = out_data.byte[0];
+
+      // also add the most recent written value to the cache
+      history_buffer_cache_val[kCacheDepth] = out_data.byte[0];
+      history_buffer_cache_idx[kCacheDepth] = history_buffer_idx;
+#pragma unroll
+      for (int j = 0; j < kCacheDepth; j++) {
+        history_buffer_cache_val[j] = history_buffer_cache_val[j + 1];
+        history_buffer_cache_idx[j] = history_buffer_cache_idx[j + 1];
+      }
+
+      // move the write index
+      history_buffer_idx =
+          (history_buffer_idx + ac_uint<1>(1)) & history_buffer_idx_mask;
+
+      // write the output to the pipe
+      OutPipe::write(OutPipeBundleT(out_data));
+    }
+  } while (!done);
+
+  OutPipe::write(OutPipeBundleT(true));
+}
+
+//
+// Top-level LZ77 decoder to switch between the single- and multi-element per

```

Comment:
```suggestion
// Top-level LZ77 decoder to select between the single- and multi-element per

```
'switch' implies to me that it's happening dynamically at runtime